cclunit-framework-source/doc/CCLUTGUIDANCE.md

```diff
@@ -0,0 +1,338 @@
+# CCL Unit Guidance
+
+## Building Unit Tests
+
+Before there can be unit testing, there must be units. The ultimate goal is to have small, testable, modular units and to test each of them individually. The goal of this guidance is to describe a way to structure a CCL script that supports testing its units individually. It will not discuss how to create units or unit testable code as there are many other sources that go into that detail.
+ 
+Say our CCL script contains a testable unit "functionX" that we want to test in some particular scenario, "scenarioY".  Then we will create and execute a unit test named something like "testFunctionXForScenarioY".  We want our test to execute functionX and the descendants of functionX. We do not want it to execute any other functions in our script. Here is how to go about it.
+ 
+First off, namespace every subroutine with the PUBLIC namespace. Then put all  (most all) of the executable code for the script into a main function, say public::main(null) = null
+Use a unique namespace to define an override for the main subroutine within the test suite, say testFunctionXForScenarioY::main(null) = null, and have the override first set up the conditions for scenarioY and then have it call functionX directly.
```

Comment:
So, I was mainly keeping this structure because this is how you worded it in your post that you gave me the link to here:
https://github.com/cerner/cclunit-framework/pull/7#discussion_r246756864

I don't want to assume on why you structured it this way (maybe thinking to show a suboptimal way first and then show a better approach?), but if I'm understanding your comment here correctly, you're wanting me to remove any namespace references within this section (Building Unit Tests), and only demonstrate the "with replace" option.  Is that accurate?  I'm trying to reconcile this with your other comment:

> This section is a distraction (from here to CCL Mocks).
It should probably be contained in a separate page if retained at all. It should certainly be delayed till after pointing out that there are two ways to mock things and describing when each is appropriate.

I'm completely fine taking out that stuff, but if I switch this earlier example to use "with replace", then some of those later statements would naturally come up here, so I want to make sure that's what you're wanting.

Reply:
Yes, that is accurate.

I learned a better way years since posting that.  I guess I didn't get around to blogging about it. 

I think it makes for a better "how to" outline if we first show how structure and call a script so that only one of its subroutines will be executed (by leveraging a main subroutine and using with replace("MAIN", alternateMain) plus your idea of a final subroutine).

After that an explanation how to mock the subroutines, scripts and CCL functions called by the subroutine being tested is in order.  Only at that point are namespaces are needed to handle the mock subroutines. 

I would be content to completely omit the section I deemed "a distraction".  In the context of a "how to" outline, it really is a distraction and the subtle points it discusses are irrelevant if the only provided instruction for mocking main is via "with replace".

Reply:
Understood.  Those changes (and your other comments) will be in the next commit.

> plus your idea of a final subroutine

I don't want to take undue credit.  It's still your idea from the blog; I was just lifting it and making minor editorial changes 😄